### DISCUSSION TITLE
Loaning Stock Items / Information for Creating an AppMixin

### DISCUSSION THREAD

**Discussion by CodingPupper3033** [inventree/InvenTree#5203] [Q&A]
Hello there,

**Overview of our needs/original issue:**
We need to mark stock items as loaned, loanable, who it was loaned to, the due date of returning the item, a front end to loan out items, amongst other things.

**What I tried:**
I have tried looking up my issue, and I cannot find any plugins that do this.
I have also tried changing the location to "Loaned", which sort of works, but it is more complicated to figure out where it originally was, who has it, etc. You can use the notes but using it in this way is more of a bodge than an intentional feature.

**Other half of the issue**
I am willing to work on a plugin for this feature. However, I need help starting. I believe the type of plugin I need to create is an AppMixin based upon the fact that I would need to add to the "central database" (have no clue how the database is setup in inventree/how to access it). How would I go about creating an AppMixin? The documentation is very limited on this. Admittedly, I do not have much experience with djangos app system, however even looking through their docs, I can't figure out how I would edit the stock item's database table, or even how to start developing an AppMixin. Despite this lack of django experience, I do have a good amount of python/js/database/general cs knowledge that once I figure out how to start, I should be able to continue from there.

**What I am asking for**
**Could I have documentation, assistance, or a crash course in starting creation of an AppMixin?** I can't figure out how to start, so I am asking here!

**Additional Info:**
I have been known to not fully explain my issues. I have tried my best here, but there is probably something obvious missing. If you need more info from me, clarification, anything, just ask; I'm not perfect.

**Comment by wolflu05** [comment]
I think you do not need the app Mixin, maybe metadata is sufficient. You can set any json as metadata on a stock location with a plugin.

**Reply by CodingPupper3033** [reply]
Thank you for your response! I appreciate the help!

This is also another good line of reasoning. With metadata, I would assume the data persists after a restart (turning off and on again), is this correct? Also, how does this metadata get stored? When I add a key value pair, in the main database will it create a new column in whatever part table?

If I were to use metadata, I do have a few more questions about that. I found the endpoint to get data for a specific part `~/api/part/43/metadata`, but not the same for only one stock item. How would I be able to do that? Also, could you link me to the documentation for the api endpoints/metadata editing endpoints? I believe I found it, but I would like to double check that there is not a "better" one.

Metadata was my other initial thought, I was just a little 'afraid' of the fact that any other plugin could just wipe the data at any time, where if I went directly to the underlying database, it would be harder (but not impossible) for another plugin to write over my data. Additionally, there is a non-zero chance we have the same key names, which though unlikely still makes me a bit timid when using the metadata.

**Reply by matmair** [reply]
If you use your plugin slug as a key it should be rather save. From a platform dev perspective writing to the database is really much more insecure.

**Reply by CodingPupper3033** [reply]
Another further question if you don't mind. How would I access the api endpoint from a plugin? I found api_caller.py in samples, and tried to replicate the `self.api_call('api/users/2')` command, yet it keeps telling me something like "https:///api/users/2" is not define. And it is not, I was assuming it would just add the url of the site to the beginning

**Reply by CodingPupper3033** [reply]
I should ask about accesing metadata, I am not thinking you might not use the api calls to access the data

**Reply by CodingPupper3033** [reply]
I solved my own question: For anyone else wondering: https://docs.inventree.org/en/latest/extend/plugins/metadata/#accessing-metadata

**Comment by matmair** [comment]
AppMixins are a very advanced feature with *no* guardrails. The plugin has full access to everything (databases, filesystem, security) and can mess up/destroy every part of the application. The lack of documentation is deliberate - at least from my side. For a start, you can look at [this plugin](https://github.com/matmair/inventree-shopify) and study [djangos tutorial](https://docs.djangoproject.com/en/4.2/intro/tutorial01/). I do not recommend using AppMixins if you are not comfortable with Django.
I do not see a reason for AppMixin usage in your description. Sounds like NavigationMixin, UrlMixin, and PanelMixin (+ maybe EventMixin) should be enough to cover it.

**Reply by CodingPupper3033** [reply]
Thanks for the help. I realize now that the concerns barring me from other mixins was unfounded and silly. Those sound like a better plan, thanks!

**Comment by CodingPupper3033** [comment]
In summary, AppMixins are an extreme measure, and one too extreme for this use case. It is suggested that you use a normal plugin and edit the metadata following this tutorial [here](https://docs.inventree.org/en/latest/extend/plugins/metadata/#accessing-metadata). 

I will hopefully start work on a plugin to managing loaning items, and share it on the inventree plugins page, so stay tuned (or create it yourself) if you need the same functionality.

Anyways, thanks so much for the help guys, really truly appreciate it!

**Comment by CodingPupper3033** [comment]
Hello again,

Since my last messages, I have gone back to believing that an AppMixin would be the best. This is because we would like to be able to see a history of lent items, have a separate set of users able to loan (ones that should not have to login to Inventree, so not able to use the default 'Users'). See what items a user is currently borrowing. Also, we would like to be able to send email notifications to users when the due date is getting closer (and passed).

If I were to do the original solution, and add Metadata to the StockItems, many operations would require iterating over every single Stock Item (Items a user is loaning and email notifications for example), which is not as good of a solution as creating new tables would be.

I have also spent some time looking through examples, learning a bit of django (obviously not an expert, but everything's making sense so far). Going through the django tutorials.

The one thing I can't seem to wrap my head around is **how do I tell Inventree/django that I have a model prepared for my plugin?** Additionally, how do I run migrations on my plugin?

Some of the things I've tried:
1. Just put a basic model file in and set my plugin to extend AppMixin  - that didn't work. Plugin Error Stack exclaims "plugins.models.X doesn't declare an explicit app_label and isn't in an application in INSTALLED_APPS."
2. Copy the files of the shopify plugin git repo (to see if there is something else besides just code you need to do for setup) - again, didn't work. I believe the issue is that I am facing is that though AppMixins add it as an installed app, it doesn't run commands to actually create the model's database and link it to my plugin.
3. Run `python manage.py runserver`, `python manage.py migrate`, `python manage.py LoaningManagementPlugin`. - All of these provide a similar error: `FileNotFoundError('INVENTREE_STATIC_ROOT not specified')` I can't find where this static root is, or if I even should as I am running this development server as a Docker.
4. In settings, click Install Plugin, type in the name of my plugin no url then click add. - Also didn't work. My guess is that either I need to provide an URL, or it couldn't find it, or I didn't make a python package, or since the migrations file is still not there.

I have tried almost every combination of my files being in a package/not in a package, file layout, imports, django manage.py commands, etc. I can't figure out how to add the model to my plugin...

**I believe the thing that has tricked me up the most is the whole AppMixin installing a plugin as an app automatically for you. I don't know how to also have it automagically add my model, or tell it to add my model, etc.**

**I understand you guys are extremely busy, so I am very appreciative of any time you are able to spend with my questions. I really appreciate y'all!**

**Reply by wolflu05** [reply]
You should be able to just create a `models.py` file next to your plugin file. It should be automatically discovered if the plugin is activated and you run the normal inventree migrations. 

To fix the `"plugins.models.X doesn't declare an explicit app_label and isn't in an application in INSTALLED_APPS."` error, you have two possibilities:
1. Name the plugin folder like your plugin slug.
2. Define the explicit `app_label` by adding a Meta class to each of your models
   ```py
   class MyModel(models.Model):
     class Meta:
       app_label = "<your-plugin-slug>"
   ```

**Reply by CodingPupper3033** [reply]
I'll definitely try this when I get back. I always had the name of the folder be the name of the plugin, and not slug. That probably was the issue.

How do I run the normal inventree migrate commands then?

Also, thanks so much for your response!

**Reply by CodingPupper3033** [reply]
Hello again,

The first method still has not worked for me. I still receive the same explicit app label issue. The second method does work, but I would like to get the first method to work. The slug is "loan" so I made a python package in the plugins folder called "loan" which hasn't worked.

**I still can't figure out how to run the "normal inventree migrations", is that done automatically?**

**Reply by SchrodingersGat** [reply]
@CodingPupper3033 run either `invoke migrate` or `invoke update` to perform migratinos

**Reply by CodingPupper3033** [reply]
> @CodingPupper3033 run either `invoke migrate` or `invoke update` to perform migratinos

Hey, thanks for the response!

Are you able to run invoke in docker/windows? Would `python -m invoke migrate' work?

Thanks again, I'm really thankful for you guys helping me get off my feet!

**Reply by SchrodingersGat** [reply]
To run these commands via docker / windows you can run the command in the active InvenTree container:

Run e.g. `docker compose exec inventree-dev-server invoke migrate`

**Reply by CodingPupper3033** [reply]
Hey, new issue, **it appears that the migration command skips over my model**. It's a subclass of django db model, the file is named model, and it isn't saying it can't figure out an app name (I used the meta class). 

I can't see any reason why it would 'skip.' When looking at it run, it does seem to discover my plugin but it doesn't migrate my models.

I don't know what to give to help yall help you. What would be helpful?

**Reply by CodingPupper3033** [reply]
How would I be able to make migrations aswell? If I'm not mistaken, this command only tells the server to attempt to migrate based upon the already made migrations. I've tried `docker compose exec inventree-dev-server invoke makemigrations` to no avail

**Comment by CodingPupper3033** [comment]
> 1. Name the plugin folder like your plugin slug.

Let's assume `SLUG = "loan"`; What should the file structure look like? I have tried all the options below, each having their own issues:

1: Issue - The other aspects of the plugin do not work (ex: nav bar)
Inventree/
├─ loan/
│  ├─ __init__.py
│  ├─ LoanPlugin.py
│  ├─ models.py
├─ plugins/
│  ├─ etc...
├─ etc...

2: Issue - Inventree seems unable to read the file? Still no navbar and things, but also, the settings page can't read the author or version (of any plugin) and produces an error at the general stage with the name frozen importlib._bootstrap, and error: `No module named 'plugins.loan'`
Inventree/
├─ plugins/
│  ├─ loan/
│  │  ├─ __init__.py
│  │  ├─ LoanPlugin.py
│  │  ├─ models.py
│  ├─ etc...
├─ etc...

Should I have the __init__.py file (should it be a package) or not? Should it be it's own "top level directory" like example 1, or a plugin like example 2?

I keep saying it, but I do keep meaning it. I appreciate you guys spending time answering common/noob questions so I may be caught up to speed. Your effort is really appreciated!

**Reply by CodingPupper3033** [reply]
markdown removed the underscore underscore from my file structure init files. Assume it's there

**Reply by wolflu05** [reply]
Actually I haven't tested 1 where I suggested it, I thought this would work, but now debugged a bit, maybe the issue is because there is no explicit AppConfig class in the AppMixin plugin architecture. The package gets added via its package path to `settings.INSTALLED_APPS` (so in your case `loan`). I think this needs some adjustment on the plugin mechanism. But @matmair maybe knows more here. A while ago where I messed around with the AppMixin I was not that familiar with Django as Im now, so I just ignored it and defined the Meta classes, it was not that hard, but I feel what you mean with getting it to work without.

Just a tip: regarding formatting, you can use markdown multi line code blocks, then its also displayed with a monospaced font.

**Reply by CodingPupper3033** [reply]
Some more testing:
Using method 2 breaks everything (all other plugins) but not as a package only breaks the plugin. But, both being a package and not being a package work if your plugin is not an app. Very very strange

**Reply by matmair** [reply]
Developing an AppMixin plugin without good prior knowledge of Django will present many challenges, I do not encourage it.

The migrations issue is part of most Django beginners problems, the solution [is simple but requires knowledge of the layout of an app in django](https://stackoverflow.com/a/56610688/17860466).

Providing support without any knowledge of the environment (you can set where plugins are located and discovered - [see docs](https://docs.inventree.org/en/latest/start/config/?h=config#plugin-options)) is fruitless. My time is limited so I chose to spend it on actually development rather then tutoring. Maybe read the plugin docs https://docs.inventree.org/en/latest/extend/how_to_plugin/ as a lot of the pitfalls you faced up to now are described there.

**Reply by CodingPupper3033** [reply]
Hello @matmair,

I understand that answering noobish questions is very frustrating; I, too, rather code than tutor; that is why I really truly appreciate any help you can give. I know it's time that you are not spent coding or anything else more useful, and I am grateful.

As for why I asked about the Django structure, I was pretty confident that you need to have it in a package inside the plugin folder, so I was confused why it wasn't working.

Thanks so much for the stack overflow question. I didn't know that making migrations does not make a folder for you. That was an assumption proved wrong.

I have looked at the plugin docs, but as someone said before, it is purposely undocumented, so I am having some difficulties getting started without a manual. Specifically, with that how to plugin page, I can not find any of my pitfalls there. Maybe I am not seeing it correctly?

I know I am asking a lot of questions, but I am trying to figure out which part of the Django tutorials applies here, how you guys set up the app, where to put things, and just all the required information/tools to get started. I again really appreciate all your help on this journey.
